File: services/learning_api/src/learning_api/routers/identity.py
"""Invitation onboarding and current learner endpoints."""

from __future__ import annotations

from fastapi import APIRouter, HTTPException

from ..dependencies import IdentityRepositoryDependency, LearnerDependency
from ..identity_contracts import (
    AcceptInvitationRequest,
    CurrentLearnerResponse,
    InvitationAcceptanceResponse,
)
from ..identity_repository import IdentityError

router = APIRouter(prefix="/api/v1", tags=["identity"])


def _safe(call):
    try:
        return call()
    except IdentityError as exc:
        raise HTTPException(
            status_code=exc.status,
            detail={"code": exc.code, "message": str(exc)},
        ) from exc


@router.post(
    "/onboarding/accept-invitation",
    operation_id="acceptBetaInvitation",
    response_model=InvitationAcceptanceResponse,
    summary="Accept a beta invitation and pin the learner to its course revision",
)
async def accept_invitation(
    body: AcceptInvitationRequest,
    learner: LearnerDependency,
    repository: IdentityRepositoryDependency,
) -> InvitationAcceptanceResponse:
    return _safe(
        lambda: repository.accept_invitation(
            learner,
            body.invitation_code,
            body.display_name,
        )
    )


@router.get(
    "/me",
    operation_id="getCurrentLearner",
    response_model=CurrentLearnerResponse,
    summary="Get the authenticated learner profile and pinned enrolments",
)
async def current_learner(
    learner: LearnerDependency,
    repository: IdentityRepositoryDependency,
) -> CurrentLearnerResponse:
    return _safe(lambda: repository.current_learner(learner))

File: apps/web/src/lib/api/identity.ts
import type { components } from "./schema";
import { createApiClient } from "./client";
import { apiErrorMessage } from "./errors";

export type CurrentLearnerResponse = components["schemas"]["CurrentLearnerResponse"];
export type InvitationAcceptanceResponse =
  components["schemas"]["InvitationAcceptanceResponse"];

export async function getCurrentLearner(
  accessToken: string,
): Promise<CurrentLearnerResponse> {
  const { data, error, response } = await createApiClient(accessToken).GET("/api/v1/me");
  if (error || !data) throw new Error(apiErrorMessage(error, response.status));
  return data;
}

export async function acceptBetaInvitation(
  accessToken: string,
  invitationCode: string,
  displayName: string,
): Promise<InvitationAcceptanceResponse> {
  const { data, error, response } = await createApiClient(accessToken).POST(
    "/api/v1/onboarding/accept-invitation",
    { body: { invitation_code: invitationCode, display_name: displayName } },
  );
  if (error || !data) throw new Error(apiErrorMessage(error, response.status));
  return data;
}
